sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelper.scala

```diff
@@ -42,15 +42,55 @@ trait ExpressionEvalHelper extends GeneratorDrivenPropertyChecks {
     InternalRow.fromSeq(values.map(CatalystTypeConverters.convertToCatalyst))
   }
 
+  protected def convertToCatalystUnsafe(a: Any): Any = a match {
+    case arr: Array[Boolean] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Byte] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Short] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Int] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Long] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Float] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case arr: Array[Double] => UnsafeArrayData.fromPrimitiveArray(arr)
+    case other => CatalystTypeConverters.convertToCatalyst(other)
+  }
+
   protected def checkEvaluation(
       expression: => Expression, expected: Any, inputRow: InternalRow = EmptyRow): Unit = {
     val serializer = new JavaSerializer(new SparkConf()).newInstance
     val expr: Expression = serializer.deserialize(serializer.serialize(expression))
-    val catalystValue = CatalystTypeConverters.convertToCatalyst(expected)
+    // No codegen version expects GenericArrayData
+    val catalystValue = expected match {
```

Comment:
Can we just ask the caller side to pass in catalyst array? We can create some helper functions at the caller side to create catalyst array easily.